eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/EventPosition.java

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.eventhubs;
+
+import com.azure.core.implementation.logging.ServiceLogger;
+import com.azure.eventhubs.implementation.AmqpConstants;
+
+import java.io.Serializable;
+import java.time.Instant;
+import java.util.Locale;
+
+import static com.azure.eventhubs.implementation.ClientConstants.END_OF_STREAM;
+import static com.azure.eventhubs.implementation.ClientConstants.START_OF_STREAM;
+
+/**
+ * Defines a position of an {@link EventData} in the event hub partition.
+ * The position can be an Offset, Sequence Number, or EnqueuedTime.
+ */
+public final class EventPosition implements Serializable {
+    private static final long serialVersionUID = 7304813338251422629L;
+
+    private final ServiceLogger logger = new ServiceLogger(EventPosition.class);
+    private final String offset;
+    private final Long sequenceNumber;
+    private final Instant dateTime;
+
+    private EventPosition(String offset, Long sequenceNumber, Instant dateTime) {
+        this.offset = offset;
+        this.sequenceNumber = sequenceNumber;
+        this.dateTime = dateTime;
+    }
+
+    /**
+     * Returns the position for the start of a stream. Provide this position in receiver creation
+     * to start receiving from the first available event in the partition.
+     *
+     * @return An {@link EventPosition} set to the start of an Event Hubs stream.
+     */
+    public static EventPosition fromStartOfStream() {
+        return new EventPosition(START_OF_STREAM, null, null);
+    }
+
+    /**
+     * Returns the position for the end of a stream. Provide this position in receiver creation
+     * to start receiving from the next available event in the partition after the receiver is created.
+     *
+     * @return An {@link EventPosition} set to the end of an Event Hubs stream.
+     */
+    public static EventPosition fromEndOfSream() {
+        return new EventPosition(END_OF_STREAM, null, null);
+    }
+
+    /**
+     * Creates a position at the given {@link Instant}.
+     *
+     * @param dateTime is the enqueued time of the event.
+     * @return An {@link EventPosition} object.
+     */
+    public static EventPosition fromEnqueuedTime(Instant dateTime) {
+        return new EventPosition(null, null, dateTime);
+    }
+
+    /**
+     * Creates a position at the given offset.
+     *
+     * @param offset is the byte offset of the event.
+     * @return An {@link EventPosition} object.
+     */
+    public static EventPosition fromOffset(String offset) {
+        return new EventPosition(offset, null, null);
+    }
```

Comment:
It would be interesting to understand this use case better. What exactly is an offset in this situation, how does the user know it, when would they use it, etc?

Reply:
For each event you receive, from Event Hubs, `EventData.SystemProperties` are populated. You can use that property to fetch the "offset" of that event. The offset corresponds to the event's position within the partition's stream.

If your receiver crashes, on resume, it could start reading from that particular offset in the stream rather than starting from the beginning.
https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-features#stream-offsets